compiler/src/iree/compiler/Dialect/Stream/Transforms/SpecializeEncodings.cpp

```diff
@@ -0,0 +1,172 @@
+
+#define DEBUG_TYPE "iree-stream-specialize-encodings"
+
+#define GEN_PASS_DEF_SPECIALIZEENCODINGSPASS
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h.inc"
+
+namespace {
+/// Returns a stably sorted list of dialect interfaces of T for all dialects
+/// used within the given module.
+template <typename T>
+SmallVector<const T *> gatherUsedDialectInterfaces(mlir::ModuleOp moduleOp) {
+  SmallPtrSet<const T *, 4> resultSet;
+  for (auto dialect : moduleOp.getContext()->getLoadedDialects()) {
+    auto *dialectInterface = dialect->getRegisteredInterface<T>();
+    if (!dialectInterface)
+      continue;
+    resultSet.insert(dialectInterface);
+  }
+
+  // NOTE: to ensure deterministic output we sort the result so that imports are
+  // always added in a consistent order.
+  SmallVector<const T *> results = {resultSet.begin(), resultSet.end()};
+  llvm::sort(
+      results, +[](const T *a, const T *b) {
+        return a->getDialect()->getNamespace().compare(
+                   b->getDialect()->getNamespace()) < 0;
+      });
+  return results;
+}
+
+// TODO(hanchung): Add "cloneWithEncoding" method to RankedTensorType.
+static RankedTensorType cloneWithEncoding(RankedTensorType type,
+                                          Attribute encoding) {
+  return RankedTensorType::get(type.getShape(), type.getElementType(),
+                               encoding);
+}
+
+static LogicalResult
+addLayoutsToTensorPhaseOps(ModuleOp moduleOp, FunctionOpInterface funcOp,
+                           LayoutAttrSolverFn makeLayoutAttrFn) {
+  SmallVector<AffinityOpInterface> candidates;
+  funcOp.walk([&](AffinityOpInterface affinityOp) {
+    // Only need to update encoding types for ops that have TensorPhaseOp trait.
+    if (!affinityOp->hasTrait<OpTrait::IREE::Stream::TensorPhaseOp>()) {
+      return;
+    }
+
+    // Bail out if the operation does not have an affinity attribute.
+    // TODO(hanchung): We should use the default device in this case. However,
+    // it is not guaranteed that default device attribute will always be set in
+    // the IR. (Is the statement correct?)
+    auto affAttr = affinityOp.getAffinityAttr();
+    if (!affAttr) {
+      return;
+    }
+    candidates.push_back(affinityOp);
+  });
+
+  if (candidates.empty()) {
+    return success();
+  }
+
+  IRRewriter rewriter(funcOp.getContext());
+  for (auto affinityOp : candidates) {
+    auto affAttr = affinityOp.getAffinityAttr();
+    SetVector<Attribute> layouts;
+    if (failed(makeLayoutAttrFn(affAttr, moduleOp, layouts))) {
+      affinityOp.emitError("failed on making layouts");
+      return failure();
+    }
+
+    auto getEncodingWithNewLayouts =
+        [=](Type type) -> std::optional<IREE::Encoding::EncodingAttr> {
+      auto rankedTensorType = dyn_cast<RankedTensorType>(type);
+      if (!rankedTensorType) {
+        return std::nullopt;
+      }
+      auto encoding = IREE::Encoding::getEncodingAttr(rankedTensorType);
```

Comment:
```suggestion
      auto encodingAttr = IREE::Encoding::getEncodingAttr(rankedTensorType);
```

consistently use the Attr suffix on attrs (especially with `auto` usage)